tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx

```diff
@@ -13,44 +13,96 @@
     */
-   std::vector<std::string> fSubTreeNames{};
+   std::vector<std::string> fTreeNames{};
+
+   /**
+    * A list of file names.
+    * They can contain the globbing characters supported by TChain. See TChain::Add for more information.
+    */
+   std::vector<std::string> fFileNameGlobs{};
+
+   ULong64_t fStartEntry{}; ///< The entry where the dataset processing should start (inclusive).
+   ULong64_t fEndEntry{};   ///< The entry where the dataset processing should end (exclusive).
+
+   ROOT::Internal::TreeUtils::RFriendInfo fFriendInfo{}; ///< List of friends
+
+   RDatasetSpec(const std::string &treeName, const std::string &fileName, REntryRange entryRange = {})
+      : fTreeNames(std::vector<std::string>{treeName}), fFileNameGlobs(std::vector<std::string>{fileName}),
+        fStartEntry(entryRange.fStartEntry), fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   RDatasetSpec(const std::string &treeName, const std::vector<std::string> &fileNames, REntryRange entryRange = {})
+      : fTreeNames(std::vector<std::string>{treeName}), fFileNameGlobs(fileNames), fStartEntry(entryRange.fStartEntry),
+        fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   RDatasetSpec(const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames,
+                REntryRange entryRange = {})
+      : fTreeNames(
+           fileNames.size() != treeNames.size() && treeNames.size() != 1
+              ? throw std::logic_error("RDatasetSpec exepcts either N trees and N files, or 1 tree and N files.")
+              : treeNames),
+        fFileNameGlobs(fileNames), fStartEntry(entryRange.fStartEntry), fEndEntry(entryRange.fEndEntry)
+   {
+   }
+
+   void AddFriend(const std::string &treeName, const std::string &fileName)
+   {
+      fFriendInfo.fFriendNames.emplace_back(
+         treeName, treeName); // TODO: user might specify alias, now it is the tree name by default
+      fFriendInfo.fFriendFileNames.emplace_back(std::vector<std::string>{fileName});
+      fFriendInfo.fFriendChainSubNames.emplace_back(std::vector<std::string>{}); // this is a tree
+   }
 
-   RDatasetSpec(const std::string &datasetName, const std::string &fileName, ULong64_t startEntry = 0,
-                ULong64_t endEntry = 0, const std::vector<std::string> &defaultColumns = {},
-                const std::vector<std::string> &subTreenames = {})
-      : fDatasetName(datasetName), fFileNameGlobs(std::vector<std::string>{fileName}), fStartEntry(startEntry),
-        fEndEntry(endEntry), fDefaultColumns(defaultColumns), fSubTreeNames(subTreenames)
+   void AddFriend(const std::string &treeName, const std::vector<std::string> &fileNames)
    {
+      fFriendInfo.fFriendNames.emplace_back(
+         treeName, treeName); // TODO: user might specify alias, now it is the tree name by default
+      fFriendInfo.fFriendFileNames.emplace_back(fileNames);
+      fFriendInfo.fFriendChainSubNames.emplace_back(std::vector<std::string>{treeName}); // now this is a chain
```

Comment:
Since this is a TChain, the current element of the `fFriendChainSubNames` vector must be a vector with the same length of the corresponding element in the `fFriendFileNames` vector.

In practice, this should be a vector with length equal to `len(fileNames)` and each element of the vector is a repetition of `treeName`.